### PR TITLE
JS Faster

### PR DESCRIPTION
**Pull request by cpcallen**
Various performance improvements that give greater overall speed than before threads were introduced.

### CODE REVIEW

File: server/interpreter.js [hunk 1]
```diff
@@ -2654,16 +2654,13 @@ Interpreter.prototype['stepCallExpression'] = function(stack, state, node) {
...
-      this.addVariableToScope(scope, paramName, state.throwValue);
-    }
-    var nextState = this.pushNode_(node['body']);
-    nextState.scope = scope;
+    // Create an empty scope.
+    var scope = new Interpreter.Scope(state.scope);
+    // Add the argument.
+    var paramName = node['param']['name'];
+    this.addVariableToScope(scope, paramName, state.throwValue);
+    // Execute catch clause
```
Comment (NeilFraser) [inline server/interpreter.js]: .

File: server/interpreter.js [hunk 2]
```diff
@@ -2654,16 +2654,13 @@ Interpreter.prototype['stepCallExpression'] = function(stack, state, node) {
...
-      // Add the argument.
-      var paramName = node['param']['name'];
-      this.addVariableToScope(scope, paramName, state.throwValue);
-    }
-    var nextState = this.pushNode_(node['body']);
-    nextState.scope = scope;
+    // Create an empty scope.
+    var scope = new Interpreter.Scope(state.scope);
+    // Add the argument.
+    var paramName = node['param']['name'];
```
Comment (NeilFraser) [inline server/interpreter.js]: Might as well inline this variable.

Comment (cpcallen) [inline server/interpreter.js]: I thought it wasn't going to all fit on one line, but fortunately it does.

File: server/interpreter.js (hunk 1 above)
Comment (cpcallen) [inline server/interpreter.js]: !

File: server/interpreter.js [hunk 3]
```diff
@@ -2846,29 +2841,31 @@ Interpreter.prototype['stepForInStatement'] = function(stack, state, node) {
...
   state.name_ = undefined;
   // Only reevaluate LHS if it wasn't a variable.
   if (state.variable_[0] !== Interpreter.SCOPE_REFERENCE) {
     state.doneVariable_ = false;
   }
+  // Sixth and finally, execute the body if there was one.
+  if (node['body']) {
+    return new Interpreter.State(node['body'], state.scope);
+  }
+  // TODO(cpcalle): in the absence of a body there is an unnecessary
```
Comment (cpcallen) [inline server/interpreter.js]: k

File: server/interpreter.js [hunk 4]
```diff
@@ -3096,17 +3091,17 @@ Interpreter.prototype['stepSwitchStatement'] = function(stack, state, node) {
...
       }
       if (state.matched_ || state.value === state.switchValue_) {
         state.matched_ = true;
         var n = state.n_ || 0;
         if (switchCase['consequent'][n]) {
           state.isSwitch = true;
-          this.pushNode_(switchCase['consequent'][n]);
+          var nextState = new Interpreter.State(
+              switchCase['consequent'][n], state.scope);
           state.n_ = n + 1;
```
Comment (cpcallen) [inline server/interpreter.js]: Not quite: the line is 81 characters.  Guess I could omit the final semicolon, seeing as it's the last statement in the block?  :-)

File: server/interpreter.js [hunk 5]
```diff
@@ -3058,7 +3054,7 @@ Interpreter.prototype['stepSequenceExpression'] = function(stack, state, node) {
   var expression = node['expressions'][n];
   if (expression) {
     state.n_ = n + 1;
-    this.pushNode_(expression);
+    return new Interpreter.State(expression, state.scope);
   } else {
```
Comment (cpcallen) [inline server/interpreter.js]: I've done a pass through all the step functions to rationalise this.  Quite a few `else`es have been removed.

**Pull request merged**